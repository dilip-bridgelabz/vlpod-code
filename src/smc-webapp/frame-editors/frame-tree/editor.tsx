/*
 *  This file is part of CoCalc: Copyright © 2020 Sagemath, Inc.
 *  License: AGPLv3 s.t. "Commons Clause" – see LICENSE.md for details
 */

import {
  React,
  rclass,
  rtypes,
  Component,
  Rendered,
  project_redux_name,
} from "../../app-framework";

import { ErrorDisplay, Loading, LoadingEstimate } from "smc-webapp/r_misc";
import { FormatBar } from "./format-bar";
import { StatusBar } from "./status-bar";
const { FrameTree } = require("./frame-tree");
import { EditorSpec, ErrorStyles } from "./types";

import { is_different, filename_extension } from "smc-util/misc2";

import { SetMap } from "./types";

import { AvailableFeatures } from "../../project_configuration";

interface FrameTreeEditorReactProps {
  name: string;
  actions: any;
  path: string;
  project_id: string;
  format_bar: boolean;
  format_bar_exclude?: SetMap;
  editor_spec: any;
}

interface FrameTreeEditorReduxProps {
  editor_settings?: Map<string, any>;
  terminal?: Map<string, any>;
  is_public: boolean;
  has_unsaved_changes: boolean;
  has_uncommitted_changes: boolean;
  read_only: boolean;
  is_loaded: boolean;
  local_view_state: Map<string, any>;
  error: string;
  errorstyle: ErrorStyles;
  cursors: Map<string, any>;
  status: string;
  load_time_estimate?: LoadingEstimate;
  value?: string;
  reload: Map<string, number>;
  resize: number; // if changes, means that frames have been resized, so may need refreshing; passed to leaf.
  misspelled_words: Set<string>;
  is_saving: boolean;
  gutter_markers: Map<string, any>;
  settings: Map<string, any>;
  complete: Map<string, any>;
  derived_file_types: Set<string>;
  available_features: AvailableFeatures;
}

type FrameTreeEditorProps = FrameTreeEditorReactProps &
  FrameTreeEditorReduxProps;

const FrameTreeEditor0 = class extends Component<FrameTreeEditorProps, {}> {
  private editor_spec: any = {};

  constructor(props) {
    super(props);
    // Copy the editor spec we will use for all future rendering
    // into our private state variable, and also do some function
    // evaluation (e.g,. if buttons is a function of the path).
    for (const type in props.editor_spec) {
      let spec = props.editor_spec[type];
      this.editor_spec[type] = spec;
    }
  }

  static reduxProps({ name, project_id }) {
    const project_store_name = project_redux_name(project_id);
    return {
      account: {
        editor_settings: rtypes.immutable.Map,
        terminal: rtypes.immutable.Map,
      },
      [name]: {
        is_public: rtypes.bool.isRequired,
        has_unsaved_changes: rtypes.bool.isRequired,
        has_uncommitted_changes: rtypes.bool.isRequired,
        read_only: rtypes.bool.isRequired,
        is_loaded: rtypes.bool.isRequired,
        local_view_state: rtypes.immutable.Map.isRequired,
        error: rtypes.string.isRequired,
        errorstyle: rtypes.string,
        cursors: rtypes.immutable.Map.isRequired,
        status: rtypes.string.isRequired,

        load_time_estimate: rtypes.immutable.Map,
        value: rtypes.string,

        reload: rtypes.immutable.Map.isRequired,
        resize: rtypes.number.isRequired, // if changes, means that frames have been resized, so may need refreshing; passed to leaf.
        misspelled_words: rtypes.immutable.Set.isRequired,
        is_saving: rtypes.bool.isRequired,

        gutter_markers: rtypes.immutable.Map.isRequired,

        settings: rtypes.immutable.Map.isRequired,

        complete: rtypes.immutable.Map.isRequired,

        derived_file_types: rtypes.immutable.Set,
      },
      [project_store_name]: {
        available_features: rtypes.immutable.Map,
      },
    };
  }

  shouldComponentUpdate(next): boolean {
    if (
      this.props.editor_settings === undefined ||
      next.editor_settings === undefined
    )
      return true;
    return (
      is_different(this.props, next, [
        // do NOT include editor_spec below -- it is assumed to never change
        "is_public",
        "has_unsaved_changes",
        "has_uncommitted_changes",
        "read_only",
        "is_loaded",
        "local_view_state",
        "error",
        "errorstyle",
        "cursors",
        "status",
        "load_time_estimate",
        "value",
        "reload",
        "resize",
        "misspelled_words",
        "has_unsaved_changes",
        "has_uncommitted_changes",
        "is_saving",
        "gutter_markers",
        "editor_settings",
        "terminal",
        "settings",
        "complete",
        "derived_file_types",
        "available_features",
      ]) ||
      this.props.editor_settings.get("extra_button_bar") !==
        next.editor_settings.get("extra_button_bar")
    );
  }

  render_format_bar(): Rendered {
    if (
      this.props.format_bar &&
      !this.props.is_public &&
      this.props.editor_settings &&
      this.props.editor_settings.get("extra_button_bar")
    )
      return (
        <FormatBar
          actions={this.props.actions}
          extension={filename_extension(this.props.path)}
          exclude={this.props.format_bar_exclude}
        />
      );
  }

  render_frame_tree(): Rendered {
    if (!this.props.is_loaded) return;
    const local = this.props.local_view_state;
    const frame_tree = local.get("frame_tree");
    const editor_state = local.get("editor_state");
    return (
      <div className={"smc-vfill"}>
        <FrameTree
          editor_spec={this.editor_spec}
          name={this.props.name}
          actions={this.props.actions}
          frame_tree={frame_tree}
          editor_state={editor_state}
          project_id={this.props.project_id}
          path={this.props.path}
          active_id={local.get("active_id")}
          full_id={local.get("full_id")}
          font_size={local.get("font_size")}
          is_only={frame_tree.get("type") !== "node"}
          cursors={this.props.cursors}
          read_only={this.props.read_only}
          is_public={this.props.is_public}
          value={this.props.value}
          reload={this.props.reload}
          resize={this.props.resize}
          misspelled_words={this.props.misspelled_words}
          has_unsaved_changes={this.props.has_unsaved_changes}
          has_uncommitted_changes={this.props.has_uncommitted_changes}
          is_saving={this.props.is_saving}
          gutter_markers={this.props.gutter_markers}
          editor_settings={this.props.editor_settings}
          terminal={this.props.terminal}
          settings={this.props.settings}
          status={this.props.status}
          complete={this.props.complete}
          derived_file_types={this.props.derived_file_types}
          available_features={this.props.available_features}
        />
      </div>
    );
  }

  render_error(): Rendered {
    if (!this.props.error) {
      return;
    }
    const style: any = {
      maxWidth: "100%",
      margin: "1ex",
      maxHeight: "30%",
      overflowY: "scroll",
    };
    if (this.props.errorstyle === "monospace") {
      style.fontFamily = "monospace";
      style.fontSize = "85%";
      style.whiteSpace = "pre-wrap";
    }
    return (
      <ErrorDisplay
        error={this.props.error}
        onClose={() => this.props.actions.set_error("")}
        style={style}
      />
    );
  }

  render_status_bar(): Rendered {
    if (!this.props.is_loaded) {
      return;
    }
    if (!this.props.status) {
      return;
    }
    return <StatusBar status={this.props.status} />;
  }

  render_loading(): Rendered {
    if (this.props.is_loaded) return;
    return (
      <div
        className="smc-vfill"
        style={{
          fontSize: "40px",
          textAlign: "center",
          padding: "15px",
          color: "#999",
        }}
      >
        <Loading estimate={this.props.load_time_estimate} />
      </div>
    );
  }

  render(): Rendered {
    return (
      <div className="smc-vfill cc-frame-tree-editor">
        {this.render_error()}
        {this.render_format_bar()}
        {this.render_loading()}
        {this.render_frame_tree()}
        {this.render_status_bar()}
      </div>
    );
  }
} as React.ComponentType<FrameTreeEditorReactProps>;

const FrameTreeEditor = rclass(FrameTreeEditor0);

interface Options {
  display_name: string;
  format_bar: boolean;
  format_bar_exclude?: SetMap;
  editor_spec: EditorSpec;
}

interface EditorProps {
  actions: any;
  name: string;
  path: string;
  project_id: string;
}

export function createEditor(opts: Options) {
  class Editor extends Component<EditorProps, {}> {
    public displayName: string = opts.display_name;

    render(): JSX.Element {
      return (
        <FrameTreeEditor
          actions={this.props.actions}
          name={this.props.name}
          path={this.props.path}
          project_id={this.props.project_id}
          format_bar={opts.format_bar}
          format_bar_exclude={opts.format_bar_exclude}
          editor_spec={opts.editor_spec}
        />
      );
    }
  }
  return Editor;
}
