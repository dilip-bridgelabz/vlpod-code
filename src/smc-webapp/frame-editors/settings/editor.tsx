/*
 *  This file is part of CoCalc: Copyright © 2020 Sagemath, Inc.
 *  License: AGPLv3 s.t. "Commons Clause" – see LICENSE.md for details
 */

/*
Settings and configuration for editing this file.
*/

import { Map } from "immutable";

import { React, Rendered, Component } from "../../app-framework";

import { is_different } from "smc-util/misc2";

// import from icon only necessary for testing via Jest
// Change to import from r_misc when it's all typescript
import { Icon } from "../../r_misc/icon";
import { Loading } from "../../r_misc/loading";

import { SpellCheck } from "./spell-check";

import { AvailableFeatures } from "../../project_configuration";

interface Props {
  id: string;
  actions: any;
  settings: Map<string, any>;
  available_features?: AvailableFeatures;
}

export class Settings extends Component<Props, {}> {
  shouldComponentUpdate(props): boolean {
    return is_different(this.props, props, ["settings", "available_features"]);
  }

  render_settings(): Rendered[] {
    const af = this.props.available_features;
    if (af == null) {
      return [<Loading key={"loading"} />];
    }
    const v: Rendered[] = [];
    this.props.settings.forEach((value, key) => {
      switch (key) {
        case "spell":
          v.push(
            <SpellCheck
              key={key}
              value={value}
              available={af.get("spellcheck")}
              set={(value) => this.props.actions.set_settings({ [key]: value })}
            />
          );
          return;
        default:
          console.warn(`UNKNOWN setting ${key} -- ignoring`);
        // we could delete it like so -- this.props.actions.set_settings({[key]:null});
      }
    });
    if (v.length == 0) {
      v.push(<div>This editor currently has no configurable settings.</div>);
    }
    return v;
  }

  render(): Rendered {
    return (
      <div
        className={"smc-vfill"}
        style={{
          overflowY: "scroll",
          padding: "5px 15px",
          fontSize: "10pt",
        }}
      >
        <h3
          style={{
            borderBottom: "1px solid #ccc",
            paddingBottom: "15px",
          }}
        >
          <Icon name="wrench" /> Editor Settings
        </h3>
        {this.render_settings()}
      </div>
    );
  }
}

export const SETTINGS_SPEC = {
  short: "Settings",
  name: "Editor Settings",
  icon: "wrench",
  buttons: {},
  component: Settings,
  hide_public: true,
};
