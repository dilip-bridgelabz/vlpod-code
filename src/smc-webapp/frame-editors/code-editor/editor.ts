/*
 *  This file is part of CoCalc: Copyright © 2020 Sagemath, Inc.
 *  License: AGPLv3 s.t. "Commons Clause" – see LICENSE.md for details
 */

/*
Top-level react component for editing code.
*/

import { CodemirrorEditor } from "./codemirror-editor";
import { filename_extension, set } from "smc-util/misc2";
import { createEditor } from "../frame-tree/editor";
import { terminal } from "../terminal-editor/editor";
import { time_travel } from "../time-travel-editor/editor";
import { file_extensions as FORMAT } from "smc-util/code-formatter";

export const SHELLS = {
  erl: "erl",
  hrl: "erl",
  py: "python3",
  sage: "sage",
  r: "R",
  m: "octave",
  jl: "julia",
  js: "node",
  ts: "ts-node",
  coffee: "coffee",
  gp: "gp",
  lua: "lua",
  ml: "ocaml",
  pl: "perl",
  rb: "ruby",
};

export const cm = {
  short: "Code",
  name: "Source Code",
  icon: "code",
  component: CodemirrorEditor,
  buttons: function (path: string): { [name: string]: true } {
    const buttons: any = set([
      "print",
      "decrease_font_size",
      "increase_font_size",
      "save",
      "time_travel",
      "replace",
      "find",
      "goto_line",
      "cut",
      "paste",
      "copy",
      "undo",
      "redo",
      "shell",
    ]);
    const ext = filename_extension(path);
    buttons.format = FORMAT.includes(ext);
    return buttons;
  },
};

const EDITOR_SPEC = {
  cm,
  terminal,
  time_travel,
};

export const Editor = createEditor({
  format_bar: false,
  editor_spec: EDITOR_SPEC,
  display_name: "CodeEditor",
});
