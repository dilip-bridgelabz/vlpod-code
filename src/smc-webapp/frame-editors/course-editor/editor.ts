/*
 *  This file is part of CoCalc: Copyright © 2020 Sagemath, Inc.
 *  License: AGPLv3 s.t. "Commons Clause" – see LICENSE.md for details
 */

/*
Spec for editing Jupyter notebooks via a frame tree.
*/

import { set } from "smc-util/misc2";
import { createEditor } from "../frame-tree/editor";
import { terminal } from "../terminal-editor/editor";
import { time_travel } from "../time-travel-editor/editor";

import {
  Assignments,
  Configuration,
  SharedProject,
  Students,
  Handouts,
} from "./course-panels";
//import { Overview } from "./overview";

const buttons = set([
  "decrease_font_size",
  "increase_font_size",
  "save",
  "time_travel" /*,
  "undo",
  "redo"*/,
]);

export const EDITOR_SPEC = {
  /*
  course_overview: {
    short: "Overview",
    name: "Course Overview",
    icon: "file",
    component: Overview,
    buttons: set([
      "decrease_font_size",
      "increase_font_size",
      "save",
      "time_travel"
    ])
  },*/
  course_students: {
    short: "Students",
    name: "Students",
    icon: "users",
    component: Students,
    buttons,
  },
  course_assignments: {
    short: "Assignments",
    name: "Assignments",
    icon: "share-square",
    component: Assignments,
    buttons,
  },
  course_handouts: {
    short: "Handouts",
    name: "Handouts",
    icon: "copy",
    component: Handouts,
    buttons,
  },
  course_configuration: {
    short: "Config",
    name: "Configuration",
    icon: "cogs",
    component: Configuration,
    buttons,
  },
  course_shared_project: {
    short: "Shared",
    name: "Shared Project",
    icon: "share-alt",
    component: SharedProject,
    buttons,
  },
  terminal,
  time_travel,
};

export const Editor = createEditor({
  format_bar: false,
  editor_spec: EDITOR_SPEC,
  display_name: "CourseEditor",
});
