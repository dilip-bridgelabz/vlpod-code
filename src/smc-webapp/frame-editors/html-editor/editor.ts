/*
 *  This file is part of CoCalc: Copyright © 2020 Sagemath, Inc.
 *  License: AGPLv3 s.t. "Commons Clause" – see LICENSE.md for details
 */

/*
Top-level react component for editing HTML documents
*/

import { createEditor } from "../frame-tree/editor";
import { set } from "smc-util/misc2";
import { QuickHTMLPreview } from "./rendered-html";
import { IFrameHTML } from "./iframe-html";
import { CodemirrorEditor } from "../code-editor/codemirror-editor";
import { SETTINGS_SPEC } from "../settings/editor";
import { terminal } from "../terminal-editor/editor";
import { time_travel } from "../time-travel-editor/editor";

const EDITOR_SPEC = {
  cm: {
    short: "Code",
    name: "Source Code",
    icon: "code",
    component: CodemirrorEditor,
    buttons: set([
      "print",
      "decrease_font_size",
      "increase_font_size",
      "save",
      "time_travel",
      "replace",
      "find",
      "goto_line",
      "cut",
      "paste",
      "copy",
      "undo",
      "redo",
      "format",
    ]),
  },

  iframe: {
    short: "HTML",
    name: "HTML IFrame",
    icon: "compass",
    component: IFrameHTML,
    buttons: set([
      "print",
      "save",
      "time_travel",
      "reload",
      "decrease_font_size",
      "increase_font_size",
    ]),
  },

  preview: {
    short: "Preview",
    name: "Quick Preview",
    icon: "html5",
    component: QuickHTMLPreview,
    buttons: set([
      "print",
      "decrease_font_size",
      "increase_font_size",
      "save",
      "time_travel",
      "reload",
    ]),
  },

  terminal,

  settings: SETTINGS_SPEC,

  time_travel,
};

export const Editor = createEditor({
  format_bar: true,
  editor_spec: EDITOR_SPEC,
  display_name: "HTMLEditor",
});
