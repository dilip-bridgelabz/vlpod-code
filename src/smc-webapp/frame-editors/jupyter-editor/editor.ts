/*
 *  This file is part of CoCalc: Copyright © 2020 Sagemath, Inc.
 *  License: AGPLv3 s.t. "Commons Clause" – see LICENSE.md for details
 */

/*
Spec for editing Jupyter notebooks via a frame tree.
*/

import { set } from "smc-util/misc2";
import { createEditor } from "../frame-tree/editor";
import { terminal } from "../terminal-editor/editor";
import { time_travel } from "../time-travel-editor/editor";

import { CellNotebook } from "./cell-notebook/cell-notebook";
import { RawIPynb } from "./raw-ipynb";

/*
import { Log } from "./log";
import { ObjectBrowser } from "./object-browser";
import { KernelConfiguration } from "./kernel-configuration";
import { Assistant } from "./assistant";
import { SingleDocNotebook } from "./singledoc-notebook";
import { MarkdownNotebook } from "./markdown-notebook";
import { DocumentationViewer } from "./documentation-viewer";
import { FileEditor } from "./file-editor";
import { PlainTerminal } from "./plain-terminal";
import { IPythonTerminal } from "./ipython-terminal";
import { Export } from "./export";
import { ClassicalNotebook } from "./classical-notebook";
*/
import { Slideshow } from "./slideshow-revealjs/slideshow";
import { TableOfContents } from "./table-of-contents/contents";
import { Introspect } from "./introspect/introspect";

export const EDITOR_SPEC = {
  jupyter_cell_notebook: {
    short: "Notebook",
    name: "Notebook (default)",
    icon: "cc-icon-ipynb",
    component: CellNotebook,
    buttons: set([
      "print",
      "decrease_font_size",
      "increase_font_size",
      "save",
      "time_travel",
      "cut",
      "paste",
      "copy",
      "undo",
      "redo",
      "format",
      "show_table_of_contents",
      /* ,
      "shell" -- disable for now since not fully implemented*/
    ]),
    customize_buttons: {
      shell: {
        text: "Console",
        title:
          "Open command line Jupyter console session attached to the same kernel as notebook",
      },
    },
  },
  jupyter_slideshow_revealjs: {
    short: "Slideshow",
    name: "Slideshow (Reveal.js)",
    icon: "slideshare",
    component: Slideshow,
    buttons: set(["build"]),
  },
  jupyter_table_of_contents: {
    short: "Contents",
    name: "Table of Contents",
    icon: "align-right",
    component: TableOfContents,
    buttons: set(["decrease_font_size", "increase_font_size"]),
  },
  introspect: {
    short: "Introspect",
    name: "Introspection",
    icon: "info",
    component: Introspect,
    buttons: set(["decrease_font_size", "increase_font_size"]),
  },
  terminal,
  time_travel,
  jupyter_raw: {
    short: "Raw",
    name: "Raw JSON editor",
    icon: "cc-icon-markdown",
    component: RawIPynb,
    buttons: set(["decrease_font_size", "increase_font_size"]),
  },
};
/*,
  jupyter_singledoc_notebook: {
    short: "SingleDoc",
    name: "Single Doc",
    icon: "code",
    component: SingleDocNotebook,
    buttons: set([])
  },
  jupyter_markdown: {
    short: "Markdown",
    name: "Markdown Doc",
    icon: "cc-icon-markdown",
    component: MarkdownNotebook,
    buttons: set([])
  },
  jupyter_log: {
    short: "Log",
    name: "Kernel Log",
    icon: "clipboard-list",
    component: Log,
    buttons: set([])
  },
  jupyter_object_browser: {
    short: "Object",
    name: "Object Browser",
    icon: "sitemap",
    component: ObjectBrowser,
    buttons: set([])
  },
  jupyter_kernel: {
    short: "Kernel",
    name: "Kernel Config",
    icon: "server",
    component: KernelConfiguration,
    buttons: set([])
  },
  jupyter_assistant: {
    short: "Assistant",
    name: "Assistant Examples",
    icon: "magic",
    component: Assistant,
    buttons: set([])
  },
  jupyter_ipython_terminal: {
    short: "IPython",
    name: "IPython Terminal",
    icon: "terminal",
    component: IPythonTerminal,
    buttons: set([])
  },
  jupyter_plain_terminal: {
    short: "Terminal",
    name: "Plain Terminal",
    icon: "terminal",
    component: PlainTerminal,
    buttons: set([])
  },
  jupyter_file_editor: {
    short: "File",
    name: "File Editor",
    icon: "cc-icon-python",
    component: FileEditor,
    buttons: set([])
  },
  jupyter_documentation_viewer: {
    short: "Docs",
    name: "Documentation",
    icon: "question-circle",
    component: DocumentationViewer,
    buttons: set([])
  },
  jupyter_export: {
    short: "Export",
    name: "Export (nbconvert)",
    icon: "file-export",
    component: Export,
    buttons: set([])
  },
  jupyter_classical: {
    short: "Classic",
    name: "Classic Notebook",
    icon: "cc-icon-ipynb",
    component: ClassicalNotebook,
    buttons: set([])
  }
*/

export const Editor = createEditor({
  format_bar: false,
  editor_spec: EDITOR_SPEC,
  display_name: "JupyterNotebook",
});
