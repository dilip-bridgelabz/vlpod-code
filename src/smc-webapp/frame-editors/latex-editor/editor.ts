/*
 *  This file is part of CoCalc: Copyright © 2020 Sagemath, Inc.
 *  License: AGPLv3 s.t. "Commons Clause" – see LICENSE.md for details
 */

/*
Spec for editing LaTeX documents.
*/

import { set } from "smc-util/misc2";

import { createEditor } from "../frame-tree/editor";

import { PDFJS } from "./pdfjs";
import { PDFEmbed } from "./pdf-embed";
import { CodemirrorEditor } from "../code-editor/codemirror-editor";
import { Build } from "./build";
import { ErrorsAndWarnings } from "./errors-and-warnings";
import { LatexWordCount } from "./latex-word-count";
import { SETTINGS_SPEC } from "../settings/editor";
import { terminal } from "../terminal-editor/editor";
import { time_travel } from "../time-travel-editor/editor";

import { pdf_path } from "./util";

export const pdfjs_buttons = set([
  "print",
  "download",
  "decrease_font_size",
  "increase_font_size",
  "zoom_page_width",
  "zoom_page_height",
  "sync",
]);

const EDITOR_SPEC = {
  cm: {
    short: "Source",
    name: "LaTeX Source Code",
    icon: "code",
    component: CodemirrorEditor,
    buttons: set([
      "build",
      "print",
      "decrease_font_size",
      "increase_font_size",
      "save",
      "time_travel",
      "replace",
      "find",
      "goto_line",
      "cut",
      "paste",
      "copy",
      "undo",
      "redo",
      "sync",
      "help",
      "format",
      "switch_to_file",
    ]),
    gutters: ["Codemirror-latex-errors"],
  },

  pdfjs_canvas: {
    short: "PDF (preview)",
    name: "PDF - Preview",
    icon: "file-pdf-o",
    component: PDFJS,
    buttons: pdfjs_buttons,
    path: pdf_path,
    style: { background: "#525659" },
    renderer: "canvas",
  },

  error: {
    short: "Errors",
    name: "Errors and Warnings",
    icon: "bug",
    component: ErrorsAndWarnings,
    buttons: set(["build"]),
  },

  build: {
    short: "Build",
    name: "Build Control and Log",
    icon: "terminal",
    component: Build,
    buttons: set([
      "build",
      "force_build",
      "clean",
      "decrease_font_size",
      "increase_font_size",
    ]),
  },

  pdf_embed: {
    short: "PDF (native)",
    name: "PDF - Native",
    icon: "file-pdf-o",
    buttons: set(["print", "save", "download"]),
    component: PDFEmbed,
    path: pdf_path,
  },

  word_count: {
    short: "Word Count",
    name: "Word Count",
    icon: "file-alt",
    buttons: set(["word_count"]),
    component: LatexWordCount,
  },

  terminal,

  settings: SETTINGS_SPEC,

  time_travel,

  /*

    latexjs: {
        short: "Preview 1",
        name: "Rough Preview  1 - LaTeX.js",
        icon: "file-pdf-o",
        component: LaTeXJS,
        buttons: set([
            "print",
            "save",
            "decrease_font_size",
            "increase_font_size"
        ])
    },

    peg: {
        short: "Preview 2",
        name: "Rough Preview 2 - PEG.js",
        icon: "file-pdf-o",
        component: PEG,
        buttons: set([
            "print",
            "save",
            "decrease_font_size",
            "increase_font_size"
        ])
    } */
};

export const Editor = createEditor({
  format_bar: true,
  format_bar_exclude: {
    strikethrough: true,
    SpecialChar: true,
    image: true,
    unformat: true,
  }, // disabled until we can properly implement them!
  editor_spec: EDITOR_SPEC,
  display_name: "LaTeXEditor",
});
