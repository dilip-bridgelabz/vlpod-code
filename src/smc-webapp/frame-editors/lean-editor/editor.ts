/*
 *  This file is part of CoCalc: Copyright © 2020 Sagemath, Inc.
 *  License: AGPLv3 s.t. "Commons Clause" – see LICENSE.md for details
 */

/*
Top-level react component for editing LEAN documents
*/

import { set } from "smc-util/misc2";
import { createEditor } from "../frame-tree/editor";
import { LeanCodemirrorEditor } from "./lean-codemirror";
import { LeanMessages } from "./lean-messages";
import { LeanInfo } from "./lean-info";
import { LeanHelp } from "./lean-help";
import { terminal } from "../terminal-editor/editor";
import { time_travel } from "../time-travel-editor/editor";

const EDITOR_SPEC = {
  "cm-lean": {
    short: "Input",
    name: "Input",
    icon: "code",
    component: LeanCodemirrorEditor,
    buttons: set([
      "print",
      "decrease_font_size",
      "increase_font_size",
      "save",
      "time_travel",
      "replace",
      "find",
      "goto_line",
      "cut",
      "paste",
      "copy",
      "undo",
      "redo",
      "restart",
      "close_and_halt",
    ]),
    gutters: ["Codemirror-lean-messages"],
  },
  "lean-info": {
    short: "Info",
    name: "Info at Cursor", // more focused -- usually used in "tactic mode"
    icon: "bullseye",
    component: LeanInfo,
    buttons: set(["decrease_font_size", "increase_font_size"]),
  },
  "lean-messages": {
    short: "Mesages",
    name: "All Messages" /* less focused -- usually used in "term mode" */,
    icon: "eye",
    component: LeanMessages,
    buttons: set(["decrease_font_size", "increase_font_size"]),
  },
  "lean-help": {
    short: "Help",
    name: "Help at Cursor",
    icon: "question-circle",
    component: LeanHelp,
    buttons: set(["decrease_font_size", "increase_font_size"]),
  },
  terminal,
  time_travel,
};

export const Editor = createEditor({
  format_bar: false,
  editor_spec: EDITOR_SPEC,
  display_name: "LeanEditor",
});
