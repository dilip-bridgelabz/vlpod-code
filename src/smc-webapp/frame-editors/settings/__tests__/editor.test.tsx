/*
 *  This file is part of CoCalc: Copyright © 2020 Sagemath, Inc.
 *  License: AGPLv3 s.t. "Commons Clause" – see LICENSE.md for details
 */

import { Settings } from "../editor";
import * as React from "react";
import { shallow } from "enzyme";
import { Map } from "immutable";
import { ALL_AVAIL, AvailableFeatures } from "../../../project_configuration";

test("renders a list of languages", () => {
  const actions: any = {};
  actions.set_settings = () => undefined;

  const settings = Map({ spell: "Value" });

  const available_features = (Map(ALL_AVAIL) as unknown) as AvailableFeatures;

  const render = shallow(
    <Settings
      actions={actions}
      settings={settings}
      available_features={available_features}
      id={"unused??"}
    />
  );

  expect(render).toMatchSnapshot();
});
