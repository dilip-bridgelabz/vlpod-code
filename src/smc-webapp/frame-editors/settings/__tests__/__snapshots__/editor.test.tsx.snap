// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a list of languages 1`] = `
<div
  className="smc-vfill"
  style={
    Object {
      "fontSize": "10pt",
      "overflowY": "scroll",
      "padding": "5px 15px",
    }
  }
>
  <h3
    style={
      Object {
        "borderBottom": "1px solid #ccc",
        "paddingBottom": "15px",
      }
    }
  >
    <Icon
      name="wrench"
    />
     Editor Settings
  </h3>
  <SpellCheck
    available={true}
    key="spell"
    set={[Function]}
    value="Value"
  />
</div>
`;
